internal/provider/resource_access_generic.go

```diff
@@ -0,0 +1,167 @@
+type genericJAASAccessResource struct {
+	client     *juju.Client
+	targetInfo resourceInfo
+
+	// subCtx is the context created with the new tflog subsystem for applications.
+	subCtx context.Context
+}
+
+// genericJAASAccessModel represents a partial generic object for access management.
+// This struct should be embedded into a struct that contains a field for a target object (normally a name or UUID).
+// Note that service accounts are treated as users but kept as a separate field for improved validation.
+type genericJAASAccessModel struct {
+	Users           types.Set    `tfsdk:"users"`
+	ServiceAccounts types.Set    `tfsdk:"service_accounts"`
+	Groups          types.Set    `tfsdk:"groups"`
+	Access          types.String `tfsdk:"access"`
+
+	// ID required by the testing framework
+	ID types.String `tfsdk:"id"`
+}
+
+func (r *genericJAASAccessResource) ConfigValidators(ctx context.Context) []resource.ConfigValidator {
+	return []resource.ConfigValidator{
+		// TODO(Kian) Add requires JAAS validator once that lands.
+		// RequiresJAASValidator{Client: r.client},
+		resourcevalidator.AtLeastOneOf(
+			path.MatchRoot("users"),
+			path.MatchRoot("groups"),
+			path.MatchRoot("service_accounts"),
+		),
+	}
+}
+
+// PartialAccessSchema returns a map of schema attributes for a JAAS access resource.
+// Access resources should use this schema and add any additional attributes e.g. name or uuid.
+func PartialAccessSchema() map[string]schema.Attribute {
+	return map[string]schema.Attribute{
+		"access": schema.StringAttribute{
+			Description: "Type of access to the model",
+			Required:    true,
+			PlanModifiers: []planmodifier.String{
+				stringplanmodifier.RequiresReplace(),
+			},
+		},
+		"users": schema.SetAttribute{
+			Description: "List of users to grant access",
+			Optional:    true,
+			ElementType: types.StringType,
+			Validators: []validator.Set{
+				setvalidator.ValueStringsAre(UsernameStringIsValid()),
+				setvalidator.ValueStringsAre(stringvalidator.RegexMatches(withAtSymbolRe, "email must contain an @ symbol")),
+			},
+		},
+		"groups": schema.SetAttribute{
+			Description: "List of groups to grant access",
+			Optional:    true,
+			ElementType: types.StringType,
+			Validators: []validator.Set{
+				setvalidator.ValueStringsAre(GroupStringIsValid()),
+			},
+		},
+		"service_accounts": schema.SetAttribute{
+			Description: "List of service account to grant access",
+			Optional:    true,
+			ElementType: types.StringType,
+			// service accounts are treated as users but defined separately
+			// for different validation and logic in the provider.
+			Validators: []validator.Set{
+				setvalidator.ValueStringsAre(UsernameStringIsValid()),
+				setvalidator.ValueStringsAre(stringvalidator.RegexMatches(noAtSymbolRe, "service accounts should not contain an @ symbol")),
+			},
+		},
+		// ID required by the testing framework
+		"id": schema.StringAttribute{
+			Computed: true,
+		},
```

Comment:
todo: delete me, only required if the resource.ResourceWithImportState interface is implemented.